tmt/steps/__init__.py

```diff
@@ -392,6 +404,136 @@ def go(self):
         self.verbose('order', self.order, 'magenta', level=3)
 
 
+class Reboot(tmt.utils.Common):
+    """ Reboot guest """
+
+    # Which kind of reboot
+    _hard = False
+
+    # Dictionary containing list of requested phases for each enabled step
+    _phases = None
+
+    # True if interactive login enabled
+    _enabled = False
+
+    def __init__(self, step, order):
+        """ Initialize relations, store the reboot order """
+        super().__init__(parent=step, name='reboot')
+        self.order = order
+
+    @classmethod
+    def command(cls, method_class=None, usage=None):
+        """ Create the reboot command """
+        @click.command()
+        @click.pass_context
+        @click.option(
+            '--hard', metavar='REBOOT_HARD', is_flag=True,
+            help='Hard reboot of provisioned machine.')
+        @click.option(
+            '--name', metavar='MACHINE_NAME', multiple=False,
+            default=None, help='Provisioned machine name.')
+        def reboot(context, **kwargs):
+            """
+            Reboot machine.
+            """
+            Reboot._save_context(context)
+            Reboot._enabled = True
+
+        return reboot
+
+    @classmethod
+    def plugins(cls, step):
+        """ Return list of reboot instances for given step """
+        if not Reboot._enabled:
+            return []
+        return [Reboot(step, phase) for phase in cls.phases(step)]
+
+    def go(self, *args, **kwargs):
+        """ Reboot the guest(s) """
+
+        # Run the interactive command
+        self.info('reboot', 'Rebooting machine', color='yellow')
+        for guest in self.parent.plan.provision.guests():
+            # Attempt to push the workdir to the guest
+            try:
+                guest.push()
+                cwd = self.parent.plan.worktree
+            except tmt.utils.GeneralError:
+                self.warn("Failed to push workdir to the guest.")
+                cwd = None
+            guest.reboot(self._hard)
+        self.info('reboot', 'Reboot finished', color='yellow')
+
+    @classmethod
+    def phases(cls, step):
```

Comment:
Since this logic is exactly the same in `login` and `reboot`, it would be nice to avoid duplication and perhaps factor out the phases handling to a new class (named something like `Action` or `ActionStep`) that both `Login` and `Reboot` would inherit from.